src/Sql/Insert.php

```diff
@@ -160,7 +160,8 @@ public function getRawState($key = null)
         $rawState = [
             'table' => $this->table,
             'columns' => array_keys($this->columns),
-            'values' => array_values($this->columns)
+            'values' => array_values($this->columns),
```

Comment:
_If anything, it seems that adding 'select' makes this whole Insert object more inspectable while keeping the Select object (and its columns) unique._
Yes, this is a major improvement already.  Many thanks for that :D :D

So part of me does agree that we can stop right here, and make an emphasis in documentation that whenever $insert->select() is used then getRawSate will not have 'columns' filled in and it will be responsibility of the API user to check for existence of 'select' key. If you go this route, I would recommend an appropriate documentation blob is added to this PR (https://github.com/zendframework/zend-db/blob/master/doc/book/sql.md#insert section) that usage of $insert->select() will make 'columns' not filled in and it will be up to the user to inspect select as you suggest for any potential conflicts manually.

Cant think of a good class to show at the moment :(

_how adding 'select' to getRawState() has any affect on Sequence other than actually making the Select object available_
This comes from API definition. Yes it may be obvious if someone reads through the source code of Zend DB prior to using it and realized that "oh, this code skips population of 'columns' in such and such conditions. Ok, I better accommodate for that". But then how many people read the source prior to usage? Then we have complains either ZF docs are incomplete or too hard to use because a developer who just wants to use the class will hit ctrl+space in editor, pick getRawState(), point it to 'columns' array and then what will he/she think when it is semi-randomly empty? Yet another bug report here. The effect it has on sequence, or any 3rd party lib inspecting the state, is something that would be covered by abstraction layer is not there.  Which leads to answer of 

_If we change the content of getRawState('columns') in this PR, then would that not be a breaking change whereas adding the ability to inspect the Select object - even though this introduces more conditionals in situations like the Sequence stuff you mention - at least gives access to what is needed without breaking how the current method/contract works_. 
If developers expect abstraction layer to do a job, in this case automatically supply list of columns that will be affected (or else whats the point of abstraction layer if I have to do things manually), then a case where that promise is broken and i have to fall back to manual inspections, **that** is breaking change to me.   If there is a common way to inspect Select object for columns, it should be included as a utility, as opposed to leaving it up to developers reinvent this part every time they call getRawState() (plus reading the source to understand why it does not follow the expectations).

I do not think you are misunderstanding as much as you say you do. (and this is not just about sequences anymore, that is just the first place I found a problem in, but API design in general) You are not giving yourself enough credit. You have good respectable arguments :) Hope I am not coming too strong against it and we are still on good discussion terms - you are giving me great thinking exercise :)

I feel we came to a standstill, both having valid arguments, and since I am not the framework's maintainer, I have no further say in right or wrong way. Will have to call to @weierophinney to be the judge.